server/task.go

```diff
@@ -891,3 +898,32 @@ func areAllTasksDone(pipeline *api.Pipeline) bool {
 	}
 	return true
 }
+
+func (s *Server) cancelDependingTasks(ctx context.Context, task *api.Task) error {
+	queue := []int{task.ID}
+	seen := map[int]bool{task.ID: true}
+	for len(queue) != 0 {
+		fromTaskID := queue[0]
+		queue = queue[1:]
+		dagList, err := s.store.FindTaskDAGList(ctx, &api.TaskDAGFind{FromTaskID: &fromTaskID})
+		if err != nil {
+			return err
+		}
+		for _, dag := range dagList {
+			if seen[dag.ToTaskID] {
+				return errors.Errorf("found a cycle in task dag, visit task %v twice", dag.ToTaskID)
+			}
+			seen[dag.ToTaskID] = true
+
+			if _, err := s.store.PatchTaskStatus(ctx, &api.TaskStatusPatch{
```

Comment:
Can we do batch patch for task status? We can do it in a follow-up PR though.

Reply:
SGTM. Will do in a follow up.